webpack/assets/javascripts/react_app/common/hooks/API/APIHooks.js

```diff
@@ -15,30 +15,30 @@ import { APIActions } from '../../../redux/API';
  * @return {object} returns an object that contains the response, status, key and 'setUrl' for setting the url dynamically
  */
 
-export const useAPI = (method, url, options) => {
+export const useAPI = (method, url, options = {}) => {
   const dispatch = useDispatch();
-  const keyRef = useRef(options?.key);
+  const { key, ...rest } = options;
+  const [keyState, setKeyState] = useState(key || uuid());
+  const [APIoptions, setAPIOptions] = useState(rest);
 
   useEffect(() => {
-    if (!keyRef.current) keyRef.current = uuid();
-  }, []);
+    if (key) setKeyState(key);
+  }, [key]);
```

Comment:
If there is a `key` option, would it ever change on a rerender of a component using this hook? It feels to me like actually `useRef` is the correct tool here. I think we could completely get rid of this `useEffect` and use
```js
const { current: keyRef } = useRef(key || uuid());
```

Reply:
I've  changed it to a state from ref just because I found an edge case - dynamic keys, let me add an example:

At the moment, host details page uses a constant key for fetching host's data, but what if we want to add a breadcrumb switcher for switching hosts?  host details page component will be re-rendered with a different pathname, wouldn't we prefer to use dynamic keys  (each host will have its corresponding state in redux):

```js
const HostDetails = ({ pathname }) => {
  // re-rendered when host is switched
  const dynamicKey = `HOST_DETAILS_${pathname}`
  const { response } = useAPI('get', hostUrl, { dynamicKey } )
}
```